### PR TITLE
Adds an API for being able to extend menus when embedding site editor

### PR DESCRIPTION
**Pull request by arjo129**
* Builds on #162 

This PR adds a simple menu for being able to extend the site editor when it is used as an embedded application. In particular it adds support for adding menu items which do not exist in the core site editor framework. This is useful for instance when say we build a custom version of site editor which exports lines in a bespoke format. Additionally, while this approach increases boiler plate, it also allows plugins to be self contained. This will be useful when scaling site editor up and lead to more readable self contained code when adding new features.

The approach used is inspired by VSCode's extension API. At startup a plugin can declare its intent to register a new menu action. When it does so it will recieve an event handle the plugin should listen for. The plugin can then listen for the event handle separately thus enabling a complete integration with the UI of the site editor without much effort.

Here is a minimal example of how the new API may be used by a downstream application:
```rust
use bevy::prelude::*;
use librmf_site_editor::ui_command::TopLevelMenuExtensions;
use librmf_site_editor::ui_command::{EventHandle, MenuEvent};
use librmf_site_editor::SiteEditor;

#[derive(Debug, Default)]
struct MyMenuPlugin;

#[derive(Debug, Default, Resource)]
struct MyMenuHandler {
    event_handler: Option<EventHandle>,
}

/// Startup system to register menu.
fn init(mut data: ResMut<TopLevelMenuExtensions>, mut menu_handle: ResMut<MyMenuHandler>) {
    // This is all it takes to register a new menu item
    // We need to keep track of the event handler in order to make
    // sure that we can check the
    menu_handle.event_handler = Some(
        data.add_item(&"My Menu".to_string(), &"My Random Action".to_string())
            .unwrap(),
    );
}

/// Handler for menu item. All one needs tp dp is check that you recieve
/// an event that is of the same type as the one we are supposed to
/// handle.
fn watch_click(mut reader: EventReader<MenuEvent>, menu_handle: Res<MyMenuHandler>) {
    let Some(ref evt) = menu_handle.event_handler else {
        return;
    };
    for event in reader.iter() {
        if event.is_same(evt) {
            println!("Custom event clicked")
        }
    }
}

/// The actual plugin
impl Plugin for MyMenuPlugin {
    fn build(&self, app: &mut App) {
        app.init_resource::<MyMenuHandler>()
            .add_startup_system(init)
            .add_system(watch_click);
    }
}

/// Lets embed site editor in our application with our own plugin
fn main() {
    App::new()
        .add_plugin(SiteEditor)
        .add_plugin(MyMenuPlugin)
        .run();
}

```
This results in the following menu:
image

Clicking on the menu will lead to a printout saying `Custom event clicked`.

Some things I had considered. It may not actually be necessary to issue a unique event handler, we could just listen for s string matching the menu item. However, I suspect this could lead to unwanted side effects where multiple plugins listen to an action that was only intended for one plugin, additionally listening for strings is more time conupming than just comparing a single integer.

Some things not considered by this:
* Order in which items appear: VS Code does provide a way of dealing with this.
* Sub-sub menus
* Visibility control
* Checkboxes and sliders

Many of these can be handled in a follow up PR.

### CODE REVIEW

**Review by mxgrey — COMMENTED** [review]
This is definitely an important feature for us to support, and in general I want the site editor to have better ways to extend the UI.

The approach of this PR (using a resource) could work, but have you considered an entity/component approach instead? The way I intended to make the UI extensible in the future was to allow users to build UI trees as entities with parent/child relationships which contain components that describe what the UI should look like. It might look something like

```
pub struct UiWidget {
    system: Arc<RwLock<dyn System<In = &mut bevy_egui::egui::Ui>, Out = ()>>>,
}
```

and then we can use queries to find these `UiWidget`s and process them during the egui rendering phase according to their parent/child relationships.

But for this PR we can just take one very modest, much simpler step in that direction. E.g.

```
#[derive(Component)]
pub struct Menu {
    text: String,
}

#[derive(Component)]
pub enum MenuItem {
    Text(String),
    Image(egui::TextureId),
    ImageAndText(egui::TextureId, String),
}
```

A `Menu` component indicates a node that can be expanded to show a list of menu items. A `MenuItem` component will show up as a button within the menu and emit an event containing its `Entity` value when it gets clicked.

A `Menu` or `MenuItem` that has no parent will show up directly in the main menu bar at the top of the screen. A `Menu` or `MenuItem` that has a `Menu` parent will show up whenever that parent `Menu` gets expanded. The `MenuItem` will be displayed as a `Button::new(~)`, `ImageButton::new(~)` or `Button::image_and_text(~, ~)` depending on which variant it has.

What are your thoughts on that approach? I could have that implemented before Monday (probably sooner if nothing else comes up). I think my proposal is more idiomatic for the way ECS is supposed to be used. One thing that stands out to me in this PR is that the `TopLevelMenuExtensions` resource slightly reinvents entities with its `EventHandle` struct, which strikes me as a good indication that we should actually be using entities for this.

**Comment by arjo129** [comment]
I'm fine with using components instead however, I was also considering the ergonomics of the API itself. It would seem that adding a menu would require us to iterate through a `Query<Menu>` and then add a `MenuItem` as a child. Perhaps a helper function could make this easier. How would we identify unique events though? Perhaps by sending an event with the entity. I can hack it together tomorrow.

I don't think we should expose the full UI system from the menu API though (it makes sense for side panels), but the main menu should be rendered by us for stylistic consistency.

**Comment by mxgrey** [comment]
> Perhaps a helper function could make this easier.

My plan for that is to use resources to store and fetch important menu nodes, e.g.:

```
/* Public API of site editor*/
#[derive(Resource)]
pub struct FileMenu(Entity);
impl FileMenu {
  pub fn get(&self) -> Entity {
    self.0
  }
}

pub struct MenuItemClicked(pub Entity);
```
```
/* Third-party user code */
#[derive(Resource)]
pub struct MyMenuItem(Entity);

impl FromWorld for MyMenuItem {
  fn from_world(world: &mut World) -> Self {
    let my_item = world.spawn(MenuItem::Text("Export".to_owned()).id();
    let file_menu = world.resource::<FileMenu>();
    world.entity_mut(file_menu.get()).add_children([my_item]);
  }
}

fn handle_my_menu_item(
  mut menu_events: EventReader<MenuItemClicked>,
  my_menu_item: Res<MyMenuItem>,
  /* ... other system params ... */
) {
  for event in menu_events.iter() {
    if event.0 == my_menu_item.0 {
      /* ... respond to click ... */
    }
  }
}
```

With this example code, the user would be adding their custom menu item under the `File` menu and responding to clicks inside of `handle_my_menu_items`.

I think it'll be possible to improve the ergonomics a bit after I've implemented `bevy_topics`, but I think this should be okay to start.

**Review by mxgrey — COMMENTED** [review]
Thanks for putting this together so quickly!

I left a few comments that are mostly style nitpicks. Otherwise this looks like exactly what I had in mind!

File: rmf_site_editor/src/widgets/menu_bar.rs [hunk 1]
```diff
@@ -17,15 +17,156 @@
...
+                    menu_items,
+                    extension_events,
+                    false,
+                );
+            }
+        });
+    } else {
+        let Ok(child_items) = children.get(*entity) else {
+        return;
+    };
```
[the remark this answers is not in the document]
Comment (luca-della-vedova) [reply rmf_site_editor/src/widgets/menu_bar.rs:144]: Random parachuting, `let else` support in rustfmt was merged [very recently](https://blog.rust-lang.org/2023/07/01/rustfmt-supports-let-else-statements.html) and will make it to 1.72 stable which is [scheduled for 24th August](https://releases.rs/docs/1.72.0/), regardless of what we try to do I'm sure we will miss something and will have to do a quick PR next week to fix CI.

**Review by mxgrey — APPROVED** [review]
Thanks for iterating on this so quickly, this looks awesome!

**Comment by arjo129** [comment]
Would appreciate a :heavy_check_mark:  on #162 as well, which this builds on.

**Pull request merged**